File: easy/205. Isomorphic Strings/solution.cs
/*
 * Solution for the 'Isomorphic Strings' problem.
 * Time complexity: O(n).
 * Space complexity: O(n).
 */

public class Solution
{
    public bool IsIsomorphic(string s, string t)
    {
        // Process:
        //   1. Check if the lengths of the strings are different.
        //      If so, return false immediately.
        //   2. Use two dictionaries to map characters:
        //      - One to map characters from 's' to 't'.
        //      - Another to map characters from 't' to 's'.
        //   3. Iterate through both strings simultaneously.
        //      - Check if the mapping for a character in 's' is consistent with its corresponding character in 't'.
        //      - Similarly, check the reverse mapping from 't' to 's'.
        //      - If any inconsistency is found, return false.
        //   4. If all character mappings are consistent, return true.

        if (s.Length != t.Length)
            return false;

        // Create dictionaries to map characters from 's' to 't' and vice versa.
        var mapS = new Dictionary<char, char>();
        var mapT = new Dictionary<char, char>();

        // Iterate through the characters of both strings.
        for (var i = 0; i < s.Length; i++)
        {
            // Check for inconsistent mappings in both dictionaries.
            if ((mapS.ContainsKey(s[i]) && mapS[s[i]] != t[i]) ||
                (mapT.ContainsKey(t[i]) && mapT[t[i]] != s[i]))
                return false;

            // Add or update the mappings in both dictionaries.
            mapS[s[i]] = t[i];
            mapT[t[i]] = s[i];
        }

        // If no inconsistencies are found, the strings are isomorphic.
        return true;
    }
}

File: easy/205. Isomorphic Strings/solution.py
# Solution for the 'Isomorphic Strings' problem.
# Time complexity: O(n), where n is the length of the strings.
# Space complexity: O(n), due to the dictionaries used to store character mappings.

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        # Process:
        #   1. Check if the lengths of the strings are different.
        #      If so, return False immediately.
        #   2. Use two dictionaries to map characters from 's' to 't' and vice versa.
        #      - Ensure that each character in 's' maps to a single, consistent character in 't'.
        #      - Similarly, ensure that each character in 't' maps to a single character in 's'.
        #   3. If any mapping is inconsistent, return False.
        #   4. Return True if all character mappings are consistent.

        if len(s) != len(t):
            return False

        # Create dictionaries to map characters from 's' to 't' and vice versa.
        map_s, map_t = {}, {}

        # Iterate through both strings simultaneously using zip.
        for ch_s, ch_t in zip(s, t):
            # Check for inconsistent mappings in either dictionary.
            if ((ch_s in map_s and map_s[ch_s] != ch_t) or
                (ch_t in map_t and map_t[ch_t] != ch_s)):
                return False
            
            # Add or update the mappings in both dictionaries.
            map_s[ch_s] = ch_t
            map_t[ch_t] = ch_s
        
        # If no inconsistencies are found, the strings are isomorphic.
        return True
